File: src/capability.ts
import type { Capability } from "./types.js";

export interface CapabilityResult {
  flags: string[];
  isWrite: boolean;
  /**
   * True when the flags carry `--force` — command execution is non-interactive.
   * Every capability now runs `--force` (read-only `ask`/`plan` need it to run
   * shell commands like `git show` without blocking on an approval prompt), so
   * the runner gates ALL of them on the deny-list, not just write caps.
   * Distinct from `isWrite`, which still gates the write-path lock and change-set.
   */
  forced: boolean;
  downgraded: boolean;
}

/**
 * Map a capability (+ the `allowUnsandboxed` second signal) to cursor-agent flags.
 * `write-unsandboxed` without `allowUnsandboxed:true` is downgraded to sandboxed `write`.
 *
 * All capabilities carry `--force`: headless (`--print`) runs have no interactive
 * approval channel, so without it `--mode ask`/`--mode plan` block forever the moment
 * the agent tries to run a shell command. `--mode plan`/`ask` still bar edits even with
 * `--force`, so read-only + `--force` = "run any read-only command unprompted".
 */
export function mapCapability(
  capability: Capability = "ask",
  allowUnsandboxed = false,
): CapabilityResult {
  switch (capability) {
    case "ask":
      return {
        flags: ["--mode", "ask", "--force"],
        isWrite: false,
        forced: true,
        downgraded: false,
      };
    case "plan":
      return {
        flags: ["--mode", "plan", "--force"],
        isWrite: false,
        forced: true,
        downgraded: false,
      };
    case "write":
      return {
        flags: ["--sandbox", "enabled", "--force"],
        isWrite: true,
        forced: true,
        downgraded: false,
      };
    case "write-unsandboxed":
      if (allowUnsandboxed) {
        return {
          flags: ["--sandbox", "disabled", "--force"],
          isWrite: true,
          forced: true,
          downgraded: false,
        };
      }
      // Missing the second signal -> downgrade to sandboxed write.
      return {
        flags: ["--sandbox", "enabled", "--force"],
        isWrite: true,
        forced: true,
        downgraded: true,
      };
  }
}

File: src/capability.rs
use crate::types::Capability;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct CapabilityResult {
    pub flags: Vec<String>,
    pub is_write: bool,
    pub forced: bool,
    pub downgraded: bool,
}

pub fn map_capability(capability: Capability, allow_unsandboxed: bool) -> CapabilityResult {
    match capability {
        Capability::Ask => CapabilityResult {
            flags: vec!["--mode".into(), "ask".into(), "--force".into()],
            is_write: false,
            forced: true,
            downgraded: false,
        },
        Capability::Plan => CapabilityResult {
            flags: vec!["--mode".into(), "plan".into(), "--force".into()],
            is_write: false,
            forced: true,
            downgraded: false,
        },
        Capability::Write => CapabilityResult {
            flags: vec!["--sandbox".into(), "enabled".into(), "--force".into()],
            is_write: true,
            forced: true,
            downgraded: false,
        },
        Capability::WriteUnsandboxed => {
            if allow_unsandboxed {
                CapabilityResult {
                    flags: vec!["--sandbox".into(), "disabled".into(), "--force".into()],
                    is_write: true,
                    forced: true,
                    downgraded: false,
                }
            } else {
                CapabilityResult {
                    flags: vec!["--sandbox".into(), "enabled".into(), "--force".into()],
                    is_write: true,
                    forced: true,
                    downgraded: true,
                }
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn ask_carries_force() {
        let r = map_capability(Capability::Ask, false);
        assert_eq!(r.flags, ["--mode", "ask", "--force"]);
        assert!(!r.is_write);
        assert!(r.forced);
        assert!(!r.downgraded);
    }

    #[test]
    fn plan_carries_force() {
        let r = map_capability(Capability::Plan, false);
        assert_eq!(r.flags, ["--mode", "plan", "--force"]);
        assert!(!r.is_write);
        assert!(r.forced);
        assert!(!r.downgraded);
    }

    #[test]
    fn write_is_sandboxed_and_forced() {
        let r = map_capability(Capability::Write, false);
        assert_eq!(r.flags, ["--sandbox", "enabled", "--force"]);
        assert!(r.is_write);
        assert!(r.forced);
        assert!(!r.downgraded);
    }

    #[test]
    fn write_unsandboxed_with_allow() {
        let r = map_capability(Capability::WriteUnsandboxed, true);
        assert_eq!(r.flags, ["--sandbox", "disabled", "--force"]);
        assert!(r.is_write);
        assert!(r.forced);
        assert!(!r.downgraded);
    }

    #[test]
    fn write_unsandboxed_without_signal_downgrades() {
        let r = map_capability(Capability::WriteUnsandboxed, false);
        assert_eq!(r.flags, ["--sandbox", "enabled", "--force"]);
        assert!(r.is_write);
        assert!(r.forced);
        assert!(r.downgraded);
    }

    #[test]
    fn default_capability_is_ask() {
        let r = map_capability(Capability::Ask, false);
        assert_eq!(r.flags, ["--mode", "ask", "--force"]);
        assert!(r.forced);
    }
}
